### PR TITLE
feat: Add optional restriction of script execution to certain object fields and values

### LINKED ISSUE #2485: Allow to restrict script execution to certain object fields
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-dashboard/issues?q=is%3Aissue).

### Current Limitation

Script execution can only be restricted per class, not per field. 

### Feature / Enhancement Description

Allow to restrict script execution to certain fields.

The syntax could be:

`<ClassName>$<FieldName>`

Not sure which delimiter is best to use. We're already using the `$` internally in Parse Server to store a Parse Object pointer in the database with the same syntax of `<ClassName>$<FieldName>`. And `$` is not allowed in a class or field name, so this may be most consistent.

### Example Use Case

For example, a purchase transaction in which there is a `buyingUser` and `sellingUser`, but only the `buyingUser` can be flagged:

```
"apps": [
  {
    "scripts": [
      {
        "title": "Flag fraudulent purchase",
        "classes": ["Transaction$buyingUser"],
        "cloudCodeFunction": "flagUser"
      }
    ]
  }
]
```

### Alternatives / Workarounds

The field can be determined server-side with `selectedField` param, to prohibit the script execution. However, for the dashboard user, it still seems possible to execute a script on a field even thought the server will refuse it.

### LINKED ISSUE #2487: Enable / disable script in context menu based on object field values
### New Feature / Enhancement Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-dashboard/issues?q=is%3Aissue).

### Current Limitation

Currently, a script is visible in the data browser cell context menu based on the object class and ([soon](https://github.com/parse-community/parse-dashboard/issues/2485#issuecomment-1612106920)) object field.

### Feature / Enhancement Description

Enable / Disable script in context menu based on a combination of object field values. The script should be shown but grayed out in the context menu, so that the dashboard user knows that a script is available but currently disabled based on field values.

### Example Use Case

For example, a purchase transaction in which there is a buyingUser and sellingUser, but only the buyingUser can be flagged:

```json
"apps": [
  {
    "scripts": [
      {
        "title": "Flag fraudulent purchase",
        "classes": ["Transaction$buyingUser"],
        "requiredObjectValues": {
           "completed": true,
           "status": 1,
           "country": "US"
        },
        "cloudCodeFunction": "flagUser"
      }
    ]
  }
]
```

The conditions should support `bool`, `string`, `number` types and conditions should be `AND` related, i.e. all conditions need to be met for a script to be enabled. The default (if no conditions are set if that no fields are required and the script is always enabled.

There may be a better syntax for how to define these conditions, especially considering possible future additions (see below).

### Possible Future Extension

- Support combinations `AND`, `OR`
- Support comparative conditions like `<`, `>=`,  `!=`

### PR DESCRIPTION
**Pull request by dblythy**
### New Pull Request Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-dashboard/issues?q=is%3Aissue).

### Issue Description

Currently, `scripts` show on all fields, and every single value.

Closes: #2487
Closes: #2485

### Approach

Adds the following script syntax:

```js
"scripts": [
    {
    "title": "Delete account",
    "classes": [
      {
        "name": "_User",
        "fields": [{ "name": "createdAt", "validator": "value => value > new Date(\"2025\")" }]
      }
    ],
    "cloudCodeFunction": "deleteAccount"
  }
]
```

If using this syntax, the `script` will only show on a field defined in `fields`, and it will be `disabled` if the `validator` returns `false`. 

### TODOs before merging

- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)

### CODE REVIEW

**Comment by mtrezza** [comment]
Wow, impressive PR!

The two referenced issues are not necessarily related. One could enable a script only on `fieldA` if `fieldB` has a specific value. Would that be possible?

**Comment by dblythy** [comment]
Updated the syntax so the full Parse Object is passed through:

```js
"validator": (row, field) => row.get(field) > new Date("2025")
```

**Review by mtrezza — COMMENTED** [review]
Could you resolve the conflicts? I've merged the lint and prettier [PR](https://github.com/parse-community/parse-dashboard/pull/2492).

**Review by mtrezza — COMMENTED** [review]
Could you also add a short example for this new feature to the README?

**Comment by mtrezza** [comment]
Btw could you see whether the prettier hook works for you? I've added husky to the repo, but not sure whether the git hook is actually working for existing clones of the repo.

**Comment by dblythy** [comment]
Didn't seem to run for me, perhaps my branch is outdated

**Comment by mtrezza** [comment]
Maybe hooks don't get committed? If I add a hook at least I don't see anything to commit. Anyway, was just a side node.

File: Parse-Dashboard/parse-dashboard-config.json [hunk 1]
```diff
@@ -1,13 +1,25 @@
 {
   "apps": [
     {
-      "serverURL": "http://localhost:1337/parse",
```
Comment (mtrezza) [inline Parse-Dashboard/parse-dashboard-config.json]: I assume you didn't intend to merge this?

File: src/components/BrowserRow/BrowserRow.react.js [hunk 2]
```diff
@@ -20,29 +20,7 @@ export default class BrowserRow extends Component {
...
-      setCopyableValue,
-      setCurrent,
-      setEditing,
-      setRelation,
-      onEditSelectedRow,
-      setContextMenu,
-      onFilterChange,
-      markRequiredFieldRow,
-    } = this.props;
+    const { className, columns, currentCol, isUnique, obj, onPointerClick, onPointerCmdClick, order, readOnlyFields, row, rowValue, rowWidth, selection, selectRow, setCopyableValue, setCurrent, setEditing, setRelation, onEditSelectedRow, setContextMenu, onFilterChange, markRequiredFieldRow } = this.props;
```
Comment (mtrezza) [inline src/components/BrowserRow/BrowserRow.react.js]: Could you run prettier and eslint? Did looks like a long line that would be refactored

**Review by mtrezza — CHANGES_REQUESTED** [review]

**Review by mtrezza — CHANGES_REQUESTED** [review]
Could you please ament the README to document this feature?

File: README.md [hunk 3]
```diff
@@ -397,6 +397,31 @@ You can specify scripts to execute Cloud Functions with the `scripts` option:
...
+      {
+        "title": "Delete account",
+        "classes": [
+          {
+            "name": "_User",
+            "fields": [
+              { "name": "createdAt", "validator": "value => value > new Date(\"2025\")" }
+            ]
+          }
+        ],
```
Comment (mtrezza) [inline README.md:416]: Now that I see this syntax side by side, it seems to be a change in in field value type. The value for key `classes` is suddenly not `Array<String>` any more but can also be `Array<Object>`. If we released the scripts feature only in alpha and/or beta version we could make this breaking change and don't have to support `Array<String>` anymore to simplify code.

**Comment by mtrezza** [comment]
@dblythy Could your take a look at the open questions? This seems almost ready to get merged.

File: README.md (hunk 3 above)
Comment (mtrezza) [reply README.md:416]: Leaving as is

**Review by mtrezza — APPROVED** [review]

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-alpha.8](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0-alpha.8)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0-beta.2](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0-beta.2)

**Comment by parseplatformorg** [comment]
🎉 This change has been released in version [6.0.0](https://github.com/parse-community/parse-dashboard/releases/tag/6.0.0)

**Pull request merged**